library/kotlin/src/io/envoyproxy/envoymobile/Client.kt

```diff
@@ -0,0 +1,73 @@
+package io.envoyproxy.envoymobile
+
+import java.nio.ByteBuffer
+
+interface Client {
+  fun startStream(request: Request, streamCallback: StreamCallback): StreamEmitter
+}
+
+interface StreamCallback {
+  fun onHeaders(statusCode: Int, headers: Map<String, List<String>>)
+  fun onData(byteBuffer: ByteBuffer, endStream: Boolean)
+  fun onMetadata(metadata: Map<String, List<String>>, endStream: Boolean)
+  fun onTrailers(trailers: Map<String, List<String>>)
+  fun onError(error: Error)
+
+  fun onCanceled() // TODO: onInterrupt()?
+  fun onCompletion()
+
+}
+
+interface StreamEmitter {
+  fun sendHeaders(headers: Map<String, List<String>>): Boolean
+  fun sendData(byteBuffer: ByteBuffer, endStream: Boolean): Boolean
```

Comment:
We can change the signature of these functions to prevent people from chaining on something that will throw:

```
fun sendData(byteBuffer: ByteBuffer): StreamEmitter
fun sendDataAndClose(byteBuffer: ByteBuffer): Void
```

Reply:
Hm, thinking about this some more. I think it's a bit strange to have `sendDataAndClose()` on this layer because we might want to send trailers/metadata also before we close. If we want to support this type of API, we'll have to support a matrix of methods which also send and close.

Would we be opposed to having a `close()` method and we can have better ergonomics which handle the matrix of possibilities at the layer on top of this?

Another option that would be a closer compromise is that we have a `fun close(trailers: Map): Void` method which will close the connection on sending trailers (that way we only support one case rather than 6 other methods)